tests/test_browser.go

```diff
@@ -0,0 +1,297 @@
+		rt:      rt,
+		state:   state,
+		http:    testServer,
+		samples: samples,
+		Browser: b,
+	}
+	if enableFileServer {
+		tb = tb.withFileServer()
+	}
+	return tb
+}
+
+// withHandle adds the given handler to the HTTP test server and makes it
+// accessible with the given pattern.
+func (b *browser) withHandle(pattern string, handler http.HandlerFunc) *browser {
+	if b.http == nil {
+		panic("You should enable HTTP test server, see: withHTTPServer option")
+	}
+	b.http.Mux.Handle(pattern, handler)
+	return b
+}
+
+const testBrowserStaticDir = "static"
+
+// withFileServer serves a file server using the HTTP test server that is
+// accessible via `testBrowserStaticDir` prefix.
+//
+// This method is for enabling the static file server after starting a test
+// browser. For early starting the file server see withFileServer function.
+func (b *browser) withFileServer() *browser {
+	const (
+		slash = string(os.PathSeparator)
+		path  = slash + testBrowserStaticDir + slash
+	)
+
+	fs := http.FileServer(http.Dir(testBrowserStaticDir))
+
+	return b.withHandle(path, http.StripPrefix(path, fs).ServeHTTP)
+}
+
+// URL returns the listening HTTP test server's URL combined with the given path.
+func (b *browser) URL(path string) string {
+	if b.http == nil {
+		panic("You should enable HTTP test server, see: withHTTPServer option")
+	}
+	return b.http.ServerHTTP.URL + path
+}
+
+// staticURL is a helper for URL("/`testBrowserStaticDir`/"+ path).
+func (b *browser) staticURL(path string) string {
+	return b.URL("/" + testBrowserStaticDir + "/" + path)
+}
+
+// attachFrame attaches the frame to the page and returns it.
+func (b *browser) attachFrame(page api.Page, frameID string, url string) api.Frame {
+	pageFn := `
+	async (frameId, url) => {
+		const frame = document.createElement('iframe');
+		frame.src = url;
+		frame.id = frameId;
+		document.body.appendChild(frame);
+		await new Promise(x => frame.onload = x);
+		return frame;
+	}
+	`
+	return page.EvaluateHandle(
+		b.rt.ToValue(pageFn),
+		b.rt.ToValue(frameID),
+		b.rt.ToValue(url)).
+		AsElement().
+		ContentFrame()
+}
+
+// detachFrame detaches the frame from the page.
+// func (b *browser) detachFrame(page api.Page, frameID string) {
+// 	pageFn := `
+// 	frameId => {
+//         	document.getElementById(frameId).remove();
+//     	}
+// 	`
+// 	page.Evaluate(
+// 		b.rt.ToValue(pageFn),
+// 		b.rt.ToValue(frameID))
+// }
+
+// launchOptions provides a way to customize browser type
+// launch options in tests.
+type launchOptions struct {
+	Debug    bool   `js:"debug"`
+	Headless bool   `js:"headless"`
+	SlowMo   string `js:"slowMo"`
+	Timeout  string `js:"timeout"`
+}
+
+// withLaunchOptions is a helper for increasing readability
+// in tests while customizing the browser type launch options.
+//
+// example:
+//
+//    b := TestBrowser(t, withLaunchOptions{
+//        SlowMo:  "100s",
+//        Timeout: "30s",
+//    })
+type withLaunchOptions = launchOptions
```

Comment:
This is a bit clunky, since it's only a type rename, and `launchOptions` could be used directly instead. Prefixing `with` doesn't really increase readability :man_shrugging:

Reply:
`with` is a name used for options by convention. You can use either of them but I prefer `withX`.